analyses/cms-open-data-ttbar/jetassignment_training.py

```diff
@@ -0,0 +1,1393 @@
+            trijet = ak.combinations(jets, 3, fields=["j1", "j2", "j3"])  # trijet candidates
+            trijet_labels = ak.combinations(labels, 3, fields=["j1", "j2", "j3"])
+            trijet["p4"] = trijet.j1 + trijet.j2 + trijet.j3  # calculate four-momentum of tri-jet system
+            trijet["label"] = trijet_labels.j1 + trijet_labels.j2 + trijet_labels.j3
+            trijet["max_btag"] = np.maximum(trijet.j1.btagCSVV2, np.maximum(trijet.j2.btagCSVV2, trijet.j3.btagCSVV2))
+            trijet = trijet[trijet.max_btag > B_TAG_THRESHOLD]  # at least one-btag in trijet candidates
+            # pick trijet candidate with largest pT and calculate mass of system
+            trijet_mass = trijet["p4"][ak.argmax(trijet.p4.pt, axis=1, keepdims=True)].mass
+            trijet_label = trijet["label"][ak.argmax(trijet.p4.pt, axis=1, keepdims=True)]
+            observable = ak.flatten(trijet_mass)
+            trijet_label = ak.flatten(trijet_label)
+            
+            # calculate features and labels
+            features, labels, which_combination = get_training_set(jets, electrons, muons, labels,
+                                                                   self.permutations_dict, self.labels_dict)
+    
+            
+        output = {"nevents": {events.metadata["dataset"]: len(events)},
+                  "features": col_accumulator(features.tolist()),
+                  "labels": col_accumulator(labels.tolist()),
+                  "observable": col_accumulator(observable.tolist()),
+                  "even": col_accumulator(even.tolist()),
+                  "trijet_label": col_accumulator(trijet_label.tolist()),}
+            
+        return output
+        
+    def postprocess(self, accumulator):
+        return accumulator
+
+# %% [markdown]
+# ### "Fileset" construction and metadata
+#
+# Here, we gather all the required information about the files we want to process: paths to the files and asociated metadata.
+
+# %% tags=[]
+fileset = utils.construct_fileset(N_FILES_MAX_PER_SAMPLE, 
+                                  use_xcache=False)
+
+# get rid of everything except ttbar__nominal for training purposes
+fileset_keys = list(fileset.keys())
+for key in fileset_keys:
+    if key!="ttbar__nominal":
+        fileset.pop(key)
+
+# %% [markdown]
+# ### Execute the data delivery pipeline
+
+# %% tags=[]
+NanoAODSchema.warn_missing_crossrefs = False
+
+if USE_DASK_PROCESSING:
+    executor = processor.DaskExecutor(client=utils.get_client(AF))
+else:
+    executor = processor.FuturesExecutor(workers=NUM_CORES)
+    
+run = processor.Runner(executor=executor, schema=NanoAODSchema, savemetrics=True, metadata_cache={}, 
+                       chunksize=CHUNKSIZE)
+
+# preprocess
+filemeta = run.preprocess(fileset, treename="Events")
+
+# process
+output, metrics = run(fileset, 
+                      "Events", 
+                      processor_instance = JetClassifier(permutations_dict, labels_dict))
```

Comment:
It would be probably useful to save the output after this step, allowing the subsequent training to be re-run from just the saved output. Having to re-run `coffea` every time might not be super convenient. We can do that in a follow-up though.